pageserver/src/tenant/timeline.rs

```diff
@@ -631,6 +633,66 @@ impl Timeline {
             return Ok(());
         }
 
+        // retry two times to allow first round to find layers which need to be downloaded, then
+        // download them, then retry compaction
+        for round in 0..ROUNDS {
+            // should we error out with the most specific error?
+            let last_round = round == ROUNDS - 1;
+
+            let res = self.compact_without_ondemand_downloads(ctx).await;
+
+            let rls = match res {
+                Ok(()) => return Ok(()),
+                Err(CompactionError::Phase1(Phase1Error::DownloadRequired(rls))) if !last_round => {
+                    // this can be done at most one time before exiting, waiting
+                    rls
+                }
+                Err(CompactionError::Phase1(e @ Phase1Error::DownloadRequired(_))) => {
+                    // it is possible that this keeps happening, because we are battling
+                    // some other process which evicts the layers we download; we should
+                    // retry quite soon anyways.
+                    return Err(e.into());
```

Comment:
would still like to see that comment in the error message and ultimately a `WARN` or `ERROR` level log message

Reply:
Good point, this will be output as `{:?}` so we currently have needlessly the whole stacktrace at least from the compaction bg task.

Is that not enough? Would like to get rid of the stacktrace though.

Reply:
Aah but because it's a thiserror::Error, the added variant will not have a stacktrace, only `anyhow::anyhow!` will have it, which means "failed to download some layers etc" will have them just as well. I should add for those specific error types to avoid having the stacktrace.

Reply:
[0790957](https://github.com/neondatabase/neon/pull/3598/commits/0790957c3ab8c4467d9b6cfabb931b8246939d68)

Reply:
It seems it might capture the stacktrace even on into coversion?

Reply:
Yes it does: https://github.com/dtolnay/anyhow/blob/546342c3bd5418734ea0313331f847fa9517fbeb/src/backtrace.rs#L48-L53

Ok so only way to avoid printing the stacktrace is to turn all compaction reporting `{:#}` or expose the CompactionError..

Reply:
Well, I'll just drop the last commit (0790957), and we'll just need to drop the feature from anyhow some other time.

But to the original point; if we have repeated downloads issue like you were asking:
> would still like to see that comment in the error message and ultimately a WARN or ERROR level log message

It would be logged by https://github.com/neondatabase/neon/blob/3569c1bacdaea7b357f927cef32cd657e82af941/pageserver/src/tenant/tasks.rs#L80

The message will look something like:

```
ts WARN Compaction failed, retrying in 2s: Repeated downloading of layers for compaction required

Stack backtrace:
   0: <T as core::convert::Into<U>>::into
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/convert/mod.rs:726:9
      pageserver::tenant::timeline::Timeline::compact::{{closure}}
             at /home/joonas/src/neon/neon-main/pageserver/src/tenant/timeline.rs:672:32
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/future/mod.rs:91:19
   1: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
```

So no logging and returning an error is required.

Reply:
Oki perhaps I missed what you wanted, guessing this with the next push:

```
ts INFO span compact requires download of 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000000169ACF0-000000000169AD69
ts INFO span download complete: <unfortunately full path to layer file>
ts INFO span Rebuilt layer map. Did 2 insertions to process a batch of 1 updates.
ts INFO span compact requires download of 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000000169AD69-00000000024DACB1
ts ERROR span Compaction failed, retrying in 2s: Compaction requires downloading multiple times (last was 1 layers), possibly battling against eviction

Stack backtrace:
```

Reply:
However, if you don't think it's worth the effort, feel free to mark this as resolved.

Reply:
Now [b59d12f](https://github.com/neondatabase/neon/pull/3598/commits/b59d12fc1bfee07627ad3fb3fd8d7a5221edbd1f) mentiones battling eviction, but doesn't include the amount of loops.

Reply:
Discussed that it's fine, had to rebase though so: [dcd322c](https://github.com/neondatabase/neon/pull/3598/commits/dcd322cb895b2b05957b4476f1ebee843a20ac7b) is fine.